lib/lhm/progress.rb

```diff
@@ -0,0 +1,202 @@
+require "lhm/printer"
+require "json"
+
+module Lhm
+  STATE_COPYING = "copying"
+  STATE_COPYING_DONE = "copying_done"
+  STATE_COPYING_FAILED = "copying_failed"
+  STATE_INITIAL = "initial"
+  STATE_SETUP_FAILED = "setup_failed"
+  STATE_SETUP_DONE = "setup_done"
+  STATE_SWITCHING_TABLES = "switching_tables"
+  STATE_SWITCHED_TABLES = "switched_tables"
+  STATE_SWITCHING_TABLES_FAILED = "switching_tables_failed"
+  STATE_TRIGGERS_DROPPED = "trigggers_dropeed"
+
+  class Progress
+    def initialize(origin, connection = nil, options: {})
+      @origin = origin
+      @connection = connection
+      @options = options
+      @throttler = options.include?(:throttler) ? options[:throttler].class.name : "None"
+      @table_switcher = options.include?(:atomic_switch) ? "AtomicSwitcher" : "LockedSwitcher"
+      @state = Lhm::STATE_INITIAL
+      @printer = options[:printer] || Printer::Percentage.new
+  
+      @rows_written = 0    
+      @speedometer_window = options[:speedometer_window] || 5 * 60
+      @speedometer = Speedometer.new(@speedometer_window)
+      @copy_speed = 0.0
+      @bytes_copied = 0
+      @avg_row_length = 0
+      @total_bytes = 0
+      @completion_percentage = 0
+      @start_time = Time.now
+
+      @retry_helper = SqlRetry.new(
+        @connection,
+        {
+          log_prefix: "Progress"
+        }.merge!(options.fetch(:retriable, {}))
+      )
+    end
+
+    attr_reader :state, :completion_percentage
+
+    def update_before_copy(start, limit)
+      populate_table_stats
+      @min_pk_key = start
+      @max_pk_key = limit
+    end
+
+    def update_during_copy(affected_rows)
+      @rows_written += affected_rows
+
+      @bytes_copied = @rows_written * @avg_row_length
+      @completion_percentage = ((@bytes_copied.to_f / @total_bytes) * 100.00).round(2)
+    
+      @speedometer << @bytes_copied
+      @copy_speed = @speedometer.speed
+
+      current_time = Time.now
+      if (current_time - @start_time > (5 * 60)) || @options.include?(:verbose_logging)
```

Comment:
We can probably notify a bit faster, since this won't actually print anything, maybe even once a minute?

Reply:
You changed the once a minute in the logs, but not here.